src/cdk/a11y/focus-monitor/focus-monitor.ts

```diff
@@ -311,7 +311,10 @@ export class FocusMonitor implements OnDestroy {
     this._setClasses(element, this._origin);
     elementInfo.subject.next(this._origin);
     this._lastFocusOrigin = this._origin;
-    this._origin = null;
+
+    // Null-out the origin after a setTimeout. This allows the full capture/bubble cycle to complete
+    // for the current event before nulling it.
+    setTimeout(() => this._origin = null);
```

Comment:
You could avoid breaking a bunch of tests by using a `Promise.resolve().then(...)` rather than the timeout. It does make it harder to flush though.

Reply:
I tried, it doesn't wait long enough. Looks like it has to be `setTimeout`